File: Medii de proiectare si programare/java/lab4/src/main/java/com/example/temampp/services/Service.java
package com.example.temampp.services;

import com.example.temampp.domains.EmployeeAtOffice;
import com.example.temampp.domains.Show;
import com.example.temampp.events.ShowChangeEvent;
import com.example.temampp.observer.Observable;
import com.example.temampp.observer.Observer;
import com.example.temampp.repositoryes.ArtistRepo;
import com.example.temampp.repositoryes.BuyerRepo;
import com.example.temampp.repositoryes.EmployeeRepo;
import com.example.temampp.repositoryes.ShowRepository;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Service implements Observable<ShowChangeEvent> {
    private final ArtistRepo artistRepo;
    private final BuyerRepo buyerRepo;
    private final EmployeeRepo employeeRepo;
    private final ShowRepository showRepository;

    private List<Observer<ShowChangeEvent>> observerList = new ArrayList<>();

    public Service(ArtistRepo artistRepo, BuyerRepo buyerRepo, EmployeeRepo employeeRepo, ShowRepository showRepository) {
        this.artistRepo = artistRepo;
        this.buyerRepo = buyerRepo;
        this.employeeRepo = employeeRepo;
        this.showRepository = showRepository;
    }

    public boolean logInEmployee(String email, String password) {
        return employeeRepo.logIn(email, password);
    }

    public List<Show> searchArtistByDate(LocalDate date) {
        return showRepository.searchArtistByDate(date);
    }


    public void sellTicketsToShow(String showName, String buyerName, Integer noOfTicket) {
        buyerRepo.sellTicketsToShow(showName, buyerName, noOfTicket);
    }

    public List<Show> getAllShows() {
        return (List<Show>) showRepository.findAll();
    }

    @Override
    public void addObserver(Observer<ShowChangeEvent> e) {
        observerList.add(e);
    }

    @Override
    public void removeObserver(Observer<ShowChangeEvent> e) {
        observerList.remove(e);
    }

    @Override
    public void notifyObservers(ShowChangeEvent t) {
        this.observerList.forEach(x -> x.update(t));
    }

    public void sellTickets(String showName, String buyerName, Integer noOFTickets) {
        buyerRepo.sellTicketsToShow(showName, buyerName, noOFTickets);
    }

    public Show findOneShow(String showName) {
        return showRepository.findOne(showName);
    }

    public Show updateShow(Show show) {
        return showRepository.update(show);
    }

    public EmployeeAtOffice findEmployeeByEmailAndPassword(String email, String password) {
        return employeeRepo.findOneByEmailAndPassword(email, password);
    }
}

File: an2/Medii de proiectare si programare/C#/tema7/server/Service.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using model.domains;
using persistence.repositoryes;
using service;
using temaCSharp.repositoryes;
using temaCSharp.repositoryes.interfaces;

namespace server
{
    public class Service : IService
    {
        private readonly IDictionary<int, IServiceObserver> _loggedEmployees;
        private readonly IRepoBuyer _repoBuyer;
        private readonly IRepoEmployee _repoEmployee;
        private readonly RepoShow _repoShow;
        private IRepository<string, Artist> _repoArtist;

        public Service(IRepository<string, Artist> repoArtist, IRepoBuyer repoBuyer, IRepoEmployee repoEmployee,
            RepoShow repoShow)
        {
            _repoArtist = repoArtist;
            _repoBuyer = repoBuyer;
            _repoEmployee = repoEmployee;
            _repoShow = repoShow;
            _loggedEmployees = new Dictionary<int, IServiceObserver>();
        }

        public void LogOut(EmployeeAtOffice employeeAtOffice, IServiceObserver client)
        {
            var localClient = _loggedEmployees[employeeAtOffice.IdEmployee];
            if (localClient == null) throw new Exception("Employee " + employeeAtOffice.Name + " is not logged in!\n");

            _loggedEmployees.Remove(employeeAtOffice.IdEmployee);
        }

        public List<Show> SearchArtistByDate(DateTime date)
        {
            return (List<Show>)_repoShow.SearchArtist(date);
        }

        public List<Show> GetAllShows(EmployeeAtOffice employeeAtOffice)
        {
            return (List<Show>)_repoShow.FindAll();
        }

        public void SellTickets(string showName, string buyerName, int noOfTickets)
        {
            Console.WriteLine("clients connected: " + _loggedEmployees.Count);
            var show = _repoShow.findOne(showName);
            if (noOfTickets < show.NoTicketsAvailable)
            {
                foreach (var iServiceObserver in _loggedEmployees)
                {
                    Console.WriteLine("Sending to: " + iServiceObserver.Key + ' ' + iServiceObserver.Value + " data: " +
                                      showName + ' ' +
                                      noOfTickets);
                    Task.Run(() =>
                    {
                        iServiceObserver.Value.TicketsSold(new Buyer(buyerName, noOfTickets, showName));
                    });
                }

                _repoBuyer.SellTicketsToShow(showName, buyerName, noOfTickets);
            }
            else
            {
                throw new Exception("Show doesn't have this many tickets!\n");
            }
        }

        public Show findOneShow(string showName)
        {
            return _repoShow.findOne(showName);
        }

        public Show UpdateShow(Show show)
        {
            return _repoShow.Update(show);
        }

        public EmployeeAtOffice FindEmployeeByEmailAndPassword(string email, string password)
        {
            return _repoEmployee.findOneByEmailAndPassword(email, password);
        }

        public void LogIn(string email, string password, IServiceObserver client)
        {
            var employeeAtOffice = _repoEmployee.findOneByEmailAndPassword(email, password);
            if (employeeAtOffice != null)
            {
                if (_loggedEmployees.ContainsKey(employeeAtOffice.IdEmployee))
                    throw new Exception("Employee already logged in!");

                _loggedEmployees[employeeAtOffice.IdEmployee] = client;
            }
            else
            {
                throw new Exception("Email or password invalid!");
            }
        }
    }
}